app/lib/aether/aether.coffee

```diff
@@ -0,0 +1,244 @@
+      throw new Error "Aether options are invalid: " + JSON.stringify(validationResults.errors, null, 4)
+
+    # Save our original options for recreating this Aether later.
+    @originalOptions = _.cloneDeep options  # TODO: slow
+
+    # Merge the given options with the defaults.
+    defaultsCopy = _.cloneDeep defaults
+    @options = _.merge defaultsCopy, options
+
+    @setLanguage @options.language
+    @allGlobals = @options.globals.concat protectBuiltins.builtinNames, Object.keys(@language.runtimeGlobals)  # After setLanguage, which can add globals.
+    #if statementStack[0]?
+    #  rng = statementStack[0].originalRange
+    #  aether.lastStatementRange = [rng.start, rng.end] if rng
+
+    Object.defineProperty @, 'lastStatementRange',
+      get: () -> 
+        rng = @esperEngine?.evaluator?.lastASTNodeProcessed?.originalRange
+        return [rng.start, rng.end] if rng
+
+  # Language can be changed after construction. (It will reset Aether's state.)
+  setLanguage: (language) ->
+    return if @language and @language.id is language
+    validationResults = optionsValidator language: language
+    unless validationResults.valid
+      throw new Error "New language is invalid: " + JSON.stringify(validationResults.errors, null, 4)
+    @originalOptions.language = @options.language = language
+    @language = new languages[language]()
+    @languageJS ?= if language is 'javascript' then @language else new languages.javascript 'ES5'
+    @reset()
+    return language
+
+  # Resets the state of Aether, readying it for a fresh transpile.
+  reset: ->
+    @problems = errors: [], warnings: [], infos: []
+    @style = {}
+    @flow = {}
+    @metrics = {}
+    @pure = null
+
+  # Convert to JSON so we can pass it across web workers and HTTP requests and store it in databases and such.
+  serialize: ->
+    _.pick @, ['originalOptions', 'raw', 'pure', 'problems', 'flow', 'metrics', 'style', 'ast']
+
+  # Convert a serialized Aether instance back from JSON.
+  @deserialize: (serialized) ->
+    aether = new Aether serialized.originalOptions
+    aether[prop] = val for prop, val of serialized when prop isnt "originalOptions"
+    aether
+
+  # Performs quick heuristics to determine whether the code will run or produce compilation errors.
+  # If thorough, it will perform detailed linting and return false if there are any lint errors.
+  canTranspile: (rawCode, thorough=false) ->
+    return true if not rawCode # blank code should compile, but bypass the other steps
+    return false if @language.obviouslyCannotTranspile rawCode
+    return true unless thorough
+    @lint(rawCode, @).errors.length is 0
+
+  # Determine whether two strings of code are significantly different.
+  # If careAboutLineNumbers, we strip trailing comments and whitespace and compare line count.
+  # If careAboutLint, we also lint and make sure lint problems are the same.
+  hasChangedSignificantly: (a, b, careAboutLineNumbers=false, careAboutLint=false) ->
```

Comment:
This method is called a lot however the implementation for various languages doesn't actually work. So it always defaults to `true`, I believe.